service/src/main/kotlin/fi/espoo/evaka/reservations/AttendanceReservationController.kt

```diff
@@ -78,120 +89,163 @@ private fun Database.Read.getUnitOperationalDays(unitId: DaycareId, dateRange: F
     .toList()
 
+// currently queried separately to be able to use existing mapper
+private fun Database.Read.getDailyServiceTimes(childIds: Set<UUID>) = createQuery(
+    """
+    SELECT *
+    FROM daily_service_time
+    WHERE child_id = ANY(:childIds)
+    """.trimIndent()
+)
+    .bind("childIds", childIds.toTypedArray())
+    .map { row -> toDailyServiceTimes(row) }
+    .toList()
+    .filterNotNull()
+    .associateBy { it.childId }
+
+private fun mapChildReservations(rows: List<UnitAttendanceReservations.QueryRow>, serviceTimes: Map<UUID, DailyServiceTimes>): List<UnitAttendanceReservations.ChildReservations> {
+    return rows
+        .groupBy { it.child }
+        .map { (child, rowsByChild) ->
+            UnitAttendanceReservations.ChildReservations(
+                child = child.copy(dailyServiceTimes = serviceTimes.get(child.id)),
+                dailyData = rowsByChild
+                    .associateBy { it.date }
+                    .map { (date, rowByDateAndChild) ->
+                        UnitAttendanceReservations.DailyChildData(
+                            date = date,
+                            reservation = rowByDateAndChild.reservation?.let {
+                                UnitAttendanceReservations.ReservationTimes(
+                                    startTime = it.startTime,
+                                    endTime = it.endTime
+                                )
+                            },
+                            attendance = rowByDateAndChild.attendance?.let {
+                                UnitAttendanceReservations.AttendanceTimes(
+                                    startTime = it.start.toLocalTime().format(DateTimeFormatter.ofPattern("HH:mm")),
+                                    endTime = it.end?.toLocalTime()?.format(DateTimeFormatter.ofPattern("HH:mm")),
+                                )
+                            },
+                            absence = rowByDateAndChild.absence?.let {
+                                UnitAttendanceReservations.Absence(
+                                    type = it.type
+                                )
+                            }
+                        )
+                    }
+                    .sortedBy { it.date }
+            )
+        }
+        .sortedBy { "${it.child.firstName} ${it.child.lastName}" }
+}
+
 data class UnitAttendanceReservations(
     val unit: String,
     val operationalDays: List<OperationalDay>,
     val groups: List<GroupAttendanceReservations>,
     val ungrouped: List<ChildReservations>
 ) {
+    data class OperationalDay(
+        val date: LocalDate,
+        val isHoliday: Boolean
+    )
+
     data class GroupAttendanceReservations(
         val group: String,
         val children: List<ChildReservations>
     )
 
     data class ChildReservations(
         val child: Child,
-        val reservations: Map<LocalDate, Reservation>
+        val dailyData: List<DailyChildData>
```

Comment:
I don't really understand why you converted this into a list.

Reply:
Hmm, it might not be necessary after all, I'll give it another try